### PR TITLE
feat(#8633) accept array on schedule.start_from

### PR DESCRIPTION
**Pull request by 1yuv**
# Description

Allow start_from on schedule to be configured using arrays as well for fixing  medic/cht-core#8633

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [x] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [x] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* __CHT_CORE_COMPOSE_URL__
* __COUCH_SINGLE_COMPOSE_URL__
* __COUCH_CLUSTER_COMPOSE_URL__

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Review by dianabarsan — COMMENTED** [review]
Great! Can you also please add an e2e test that has a schedule with multiple fields?

File: shared-libs/transitions/src/lib/schedules.js [hunk 1]
```diff
@@ -97,8 +97,8 @@ module.exports = {
     if (typeof startFrom === 'undefined') {
       startFrom = 'reported_date';
     }
-
-    const docStart = objectPath.get(doc, startFrom);
+    
+    const docStart = Array.isArray(startFrom)? objectPath.coalesce(doc, startFrom): objectPath.get(doc, startFrom);
```
[the remark this answers is not in the document]
Comment (1yuv) [reply shared-libs/transitions/src/lib/schedules.js:101]: coalesce expectes input to be an array and will [fail](https://github.com/mariocasciaro/object-path/blob/master/index.js#L242-L245) if string parameter is passed as of object-path 0.11.8.

[the remark this answers is not in the document]
Comment (dianabarsan) [reply shared-libs/transitions/src/lib/schedules.js:101]: uf, thanks for checking.

**Review by garethbowen — COMMENTED** [review]
Awesome!

I think there are other ways to do it but this is a nice simple solution that solves the problem today. If you can add the e2e test I think we can call this done.

File: shared-libs/transitions/src/lib/schedules.js [hunk 2]
```diff
@@ -97,8 +97,8 @@ module.exports = {
     if (typeof startFrom === 'undefined') {
       startFrom = 'reported_date';
     }
-
-    const docStart = objectPath.get(doc, startFrom);
+
```
Comment (garethbowen) [inline shared-libs/transitions/src/lib/schedules.js:100]: Unnecessary whitespace...

```suggestion

```

File: shared-libs/transitions/test/integration/schedules.js [hunk 3]
```diff
@@ -732,6 +732,132 @@ describe('functional schedules', () => {
...
+      from: contact.phone,
+      contact: contact
+    };
+
+    return transition.onMatch({ doc: doc }).then(complete => {
+      assert.equal(complete, true);
+      assert(doc.tasks);
+      assert.equal(doc.tasks && doc.tasks.length, 1);
+      assert(doc.scheduled_tasks);
+      assert.equal(doc.scheduled_tasks && doc.scheduled_tasks.length, 1);
```
Comment (garethbowen) [inline shared-libs/transitions/test/integration/schedules.js:788]: I think you can use the fancy new optional chaining operator here...

```suggestion
      assert.equal(doc.tasks?.length, 1);
      assert(doc.scheduled_tasks);
      assert.equal(doc.scheduled_tasks?.length, 1);
```

**Review by garethbowen — APPROVED** [review]
A couple of nitpicks inline but I thin we're basically there. Thanks!

File: tests/integration/sentinel/schedules/schedules-start-from.spec.js [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+        chai.expect(updWithClinic.scheduled_tasks).to.be.ok;
+        chai.expect(updWithClinic.scheduled_tasks).to.have.lengthOf(3);
+
+        chai.expect(updWithClinic.scheduled_tasks[0]).to.deep.nested.include(expectedMessage2('scheduled'));
+        chai.expect(updWithClinic.scheduled_tasks[1]).to.deep.nested.include(expectedMessage3('scheduled'));
+        chai.expect(updWithClinic.scheduled_tasks[2]).to.deep.nested.include(expectedMessage4('scheduled'));
+      });
+  });
+});
```
Comment (dianabarsan) [inline tests/integration/sentinel/schedules/schedules-start-from.spec.js:253]: I think we'll need to assert on the due date as well, to make sure that the correct date is picked, not just that the message gets created.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Great work, loving the e2e test, thanks a lot for adding it. I just have one small request inline.

File: tests/integration/sentinel/schedules/schedules-start-from.spec.js (hunk 4 above)
Comment (1yuv) [reply tests/integration/sentinel/schedules/schedules-start-from.spec.js:253]: I've verified that by calculating difference between duedate and start_date and ensuring it matches our offset [here](https://github.com/medic/cht-core/pull/8639/files/aa30f4ed972750cb45adef30fe9a20f2d1edc6e5#diff-223091154c27112e2c6c6930b63c353c7ab2799567fb69f49546fffcccc5a384R243)

Comment (dianabarsan) [reply tests/integration/sentinel/schedules/schedules-start-from.spec.js:253]: Yea, I mean assert on every scheduled task, maybe included the expectedMessage function?

File: shared-libs/transitions/test/integration/schedules.js (hunk 3 above)
Comment (1yuv) [reply shared-libs/transitions/test/integration/schedules.js]: Thanks, addressed.

**Review by dianabarsan — APPROVED** [review]
Awesome! Thank you so much for adding an e2e test, I know schedules were not covered at all so this is great addition to see.

File: tests/integration/sentinel/schedules/schedules-start-from.spec.js
Comment (lorerod) [inline tests/integration/sentinel/schedules/schedules-start-from.spec.js:1]: I was thinking about edge cases or what could go wrong. What happens if we define start_from as an array of more than two elements? What happens if every field in the start_from array is unavailable?
Does this make sense?

**Review by lorerod — APPROVED** [review]
Thank you, @1yuv, for this change and for adding an integration test to cover the change. This PR looks excellent!
I left a comment if it makes sense to implement an edge case. But it is definitely not a blocker. 
I couldn't find any issue on cht-docs to update the schedule documentation and include this new feature. I can create one and work with you on the change.
Let me know what you think.

File: tests/integration/sentinel/schedules/schedules-start-from.spec.js
Comment (1yuv) [reply tests/integration/sentinel/schedules/schedules-start-from.spec.js:1]: Hi @lorerod , sure we can. Some of these were already in integraiton tests. I've added more tests and also added comment and made variable names more explanatory.

**Comment by 1yuv** [comment]
Hi @lorerod , medic/cht-docs#1207 is the releated issue for documentation update and has been approved on medic/cht-docs#1208. Also, this change required cht-conf to be updated. Related issue for cht-conf was created [here](https://github.com/medic/cht-conf/issues/578) and fixed by this [PR](https://github.com/medic/cht-conf/pull/579).
